### ISSUE TITLE
CFSecurity - If there are no assets available during deposit, the user will receive fewer tokens than they deserve when withdrawing

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-06-union-finance-update-2-judging#106]
CFSecurity

High

# If there are no assets available during deposit, the user will receive fewer tokens than they deserve when withdrawing

### Summary

The vulnerability occurs in the token staking/unstaking proces. This vulnerability can lead to users receiving fewer tokens than they should when unstaking, resulting in a loss of funds. The issue arises from a combination of failed deposits due to the disable of the market from the 3rd party market contract and the way the unstaking process handles token withdrawals from different sources.

### Root Cause

The root cause of this vulnerability is a logical error in the calculation of the `remaining` value during the unstaking process, specifically in the [`AssetManager::withdraw()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) function. When there are tokens in the contract that were not successfully deposited (due to market conditions, which can be caused by the 3rd party market contract), the function incorrectly reduces the `remaining` value after transferring these tokens.

This leads to an incorrect calculation of the remaining tokens to be withdrawn, resulting in users receiving fewer tokens than they should.

### Internal pre-conditions

None

### External pre-conditions

None

### Attack Path

1. User calls [`UserManager::stake()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/user/UserManager.sol#L738) with 5e18 tokens, but the deposit fails to register in the supported market because is disabled for some reason.

2. The 5e18 tokens remain in the contract, available for withdrawal.

3. User has a previous successful stake of 10e18 tokens.

4. User initiates an unstake of their total balance (10e18 tokens) by calling [`UserManager::unstake()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/user/UserManager.sol#L767).

5. [`AssetManager::withdraw()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) is then called to process the token withdrawal.

6. In [`AssetManager::withdraw()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335), the function first sets `remaining` to `amount`(10e18), after that it checks if there are tokens directly in the contract (selfBalance). It finds 5e18 tokens:

```solidity
uint256 remaining = amount; //10e18

// If there are tokens in Asset Manager then transfer them on priority
uint256 selfBalance = IERC20Upgradeable(token).balanceOf(address(this));
```

7. As `selfBalance` is > 0 and < `remaining` the function sets `withdrawAmount` to `selfBalance`(5e18):

```solidity
uint256 selfBalance = IERC20Upgradeable(token).balanceOf(address(this));
        if (selfBalance > 0) {
            uint256 withdrawAmount = selfBalance < remaining ? selfBalance : remaining;
```

8. `withdrawAmount`(5e18) is substracted from `remaining`, so now remaining is 5e18 instead of 10e18:

```solidity
remaining -= withdrawAmount; 
```

9. The withdrawAmount(5e18) is transferred to the user:

```solidity
IERC20Upgradeable(token).safeTransfer(account, withdrawAmount);
```

10. The function then checks if the market is supported and calculates the available supply:

```solidity
if (isMarketSupported(token)) {
            uint256 withdrawSeqLength = withdrawSeq.length;
            // iterate markets according to defined sequence and withdraw
            for (uint256 i = 0; i < withdrawSeqLength && remaining > 0; i++) {
                IMoneyMarketAdapter moneyMarket = withdrawSeq[i];
                if (!moneyMarket.supportsToken(token)) continue;

                uint256 supply = moneyMarket.getSupply(token);
```

11.  As `remaining` is < `supply`, `withdrawAmount` is set to `remaining`, which is now 5e18 due to the incorrect reduction:

```solidity
uint256 withdrawAmount = supply < remaining ? supply : remaining;
```

12. The final 5e18 tokens are withdrawn from the supported market and transferred to the user:

```solidity
if (moneyMarket.withdraw(token, account, withdrawAmount)) {
                    remaining -= withdrawAmount;
                }
```

13. [`UserManager::unstake()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/user/UserManager.sol#L767) completes, having transferred only 10e18 tokens to the user instead of the expected 15e18.

This attack path demonstrates how the specific function calls and their internal logic lead to the vulnerability, resulting in the user receiving fewer tokens than they should during the unstaking process.

### Impact

The flaw in the  [`AssetManager::withdraw()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) function, leads to an incorrect calculation of the remaining tokens to be withdrawn, resulting in users receiving fewer tokens than they should.

### PoC

Check **Attack Path**

### Mitigation

Don't substract `withdrawAmount` from `remaining` in the [`AssetManager::withdraw()`](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) function.

**Comment by cholakovvv** [comment]
Escalate 

With all due respect, I believe your understanding is incorrect. If the code had worked correctly with the proper withdrawal design, the user should have received 15e18 tokens as per the example in the report. Instead, they received 10e18 tokens, resulting in a loss of 5e18 tokens due to the explained issue.

Please check it again carefully, as this issue causes users to lose a significant portion of their deserved reward.

**Comment by sherlock-admin3** [comment]
> Escalate 
> 
> With all due respect, I believe your understanding is incorrect. If the code had worked correctly with the proper withdrawal design, the user should have received 15e18 tokens as per the example in the report. Instead, they received 10e18 tokens, resulting in a loss of 5e18 tokens due to the explained issue.
> 
> Please check it again carefully, as this issue causes users to lose a significant portion of their deserved reward.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by mystery0x** [comment]
`remaining` is added back to `balances[msg.sender][token]`:

https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L370

```
            balances[msg.sender][token] = balances[msg.sender][token] - amount + remaining;
```
So `msg.sender` can always withdraw `remaining` in the future. It does not result in a loss of funds.

**Comment by cholakovvv** [comment]
The problem here is that although remaining is added back to `balances[msg.sender][token]`, `balances` is not added anywhere within the [AssetManager::withdraw()](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) function to the amount the user should withdraw. This means that the user will lose the `remaining` and cannot withdraw it in the future.

This is because `balances` is meant to record the admin or UserManager balance, not the AssetManager balance:
https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L38C4-L42C69
```solidity
 /**
     * @dev Record admin or userManager balance
     * @dev Maps user to token to balance
     */
    mapping(address => mapping(address => uint256)) public balances;
```

Considering all these points, the issue should be validated as High, given the possible loss of funds.

**Comment by WangSecurity** [comment]
I don't understand one part of the attack path:
1. In Step 1, the user deposits (via `stake`) 5e18 tokens. Hence, in the `stake` function [L747](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/7ffe43f68a1b8e8de1dfd9de5a4d89c90fd6f710/union-v2-contracts/contracts/user/UserManager.sol#L747) their funds are added to the Staker struct, even though they're not actually deposited.
2. In Step 3 we know the user has 10e18 already staked. Hence, the user has 15e18 staked.
3. In step 4 the user initiates the unstaking process with only 10e18 tokens.

The question is, why the user should get 15e18 in the end, if they initially decided to withdraw/unstake only 10e18? I understand I may be missing something, so please correct me.

**Comment by cholakovvv** [comment]
Hey @WangSecurity, thanks for taking the time to review this!

1. In step 1 the staking flow is as follows:

- The user calls [UserManager::stake()](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/user/UserManager.sol#L738) with 5e18 tokens. This calls AssetManager::deposit(), where the 5e18 tokens are transferred to the contract. However, because the market is not supported at the moment, it does not enter [this if condition](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/7ffe43f68a1b8e8de1dfd9de5a4d89c90fd6f710/union-v2-contracts/contracts/asset/AssetManager.sol#L288) where if the tokens were successfully deposited, they would be transferred to the current moneyMarket, but this is not the case (this is the intended flow if the market is not supported). After that the function returns true. Therefore, we have 5e18 tokens in the AssetManager contract.

2. In step 3 the user has 10e18 succesfully staked tokens which he has previously staked yes, but the total amount of staked tokens is 10e18 not 15e18, because the 5e18 tokens stake failed and they're just in the AssetManager contract.

3.In step 4, the user initiates unstake() with 10e18, as this is the amount he has successfully staked. He expects his failed 5e18 deposit to be returned along with the 10e18, as indicated by the [comment in the withdraw() function](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/7ffe43f68a1b8e8de1dfd9de5a4d89c90fd6f710/union-v2-contracts/contracts/asset/AssetManager.sol#L344C9-L344C79):
>// If there are tokens in Asset Manager then transfer them on priority

So the 5e18 tokens in the AssetManager contract should be transferred to the user in addition to the 10e18.

4. This is why the user should receive 15e18 tokens, but he only receives 10e18 due to the issue mentioned in the **Root Cause** section of the report:
>The root cause of this vulnerability is a logical error in the calculation of the remaining value during the unstaking process, specifically in the [AssetManager::withdraw()](https://github.com/sherlock-audit/2024-06-union-finance-update-2/blob/main/union-v2-contracts/contracts/asset/AssetManager.sol#L335) function. When there are tokens in the contract that were not successfully deposited (due to market conditions, which can be caused by the 3rd party market contract), the function incorrectly reduces the remaining value after transferring these tokens.
This leads to an incorrect calculation of the remaining tokens to be withdrawn, resulting in users receiving fewer tokens than they should.

And explained more deeply in the **Attack Path**, which you can check.

Hope you understanded now why the user should get 15e18 tokens, but he receives just 10e18.

**Comment by WangSecurity** [comment]
> And explained more deeply in the Attack Path, which you can check

Yep, I've checked the attack path before writing the previous comment.

> Hope you understanded now why the user should get 15e18 tokens, but he receives just 10e18

I'm sorry but I don't actually. The question I've got, if in step 3, the user calls `unstake` with 15e18.

Then, in step 8:

> withdrawAmount(5e18) is substracted from remaining, so now remaining is 5e18 instead of 10e18

withdrawAmount is still 5e18. It's subtracted from the remaining, now remaining is 10e18 instead of 15e18.

Then, I assume it should revert at some point, but I'm not sure where and should go through, since 5e18 are already sent from the AssetManager directly and the other 10e18 should be withdrawn from the underlying market since they weren't withdrawn yet.

I think it would be better if you could make a coded POC here so we don't have to go back and forth clarifying small parts.

**Comment by WangSecurity** [comment]
> And explained more deeply in the Attack Path, which you can check

Yep, I've checked the attack path before writing the previous comment.

> Hope you understanded now why the user should get 15e18 tokens, but he receives just 10e18

I'm sorry but I don't actually. The question I've got, if in step 3, the user calls `unstake` with 15e18.

Then, in step 8:

> withdrawAmount(5e18) is substracted from remaining, so now remaining is 5e18 instead of 10e18

withdrawAmount is still 5e18. It's subtracted from the remaining, now remaining is 10e18 instead of 15e18.

Then, I assume it should revert at some point, but I'm not sure where and should go through, since 5e18 are already sent from the AssetManager directly and the other 10e18 should be withdrawn from the underlying market since they weren't withdrawn yet.

I think it would be better if you could make a coded POC here so we don't have to go back and forth clarifying small parts.

**Comment by cholakovvv** [comment]
Ok, in that case, I'll send later the coded PoC

**Comment by cholakovvv** [comment]
@WangSecurity here is the coded PoC:

Steps to run it:

1. Modify the TestUserManagerBase.sol like this:

```solidity
pragma solidity ^0.8.0;

import {TestWrapper} from "../TestWrapper.sol";
import {UserManager} from "union-v2-contracts/user/UserManager.sol";
import {AssetManagerMock} from "union-v2-contracts/mocks/AssetManagerMock.sol";
import {UnionTokenMock} from "union-v2-contracts/mocks/UnionTokenMock.sol";
import {FaucetERC20} from "union-v2-contracts/mocks/FaucetERC20.sol";
import {ComptrollerMock} from "union-v2-contracts/mocks/ComptrollerMock.sol";
import {UTokenMock} from "union-v2-contracts/mocks/UTokenMock.sol";
import {UserManagerOp} from "union-v2-contracts/user/UserManagerOp.sol";

import {AssetManager} from "union-v2-contracts/asset/AssetManager.sol";

contract TestUserManagerBase is TestWrapper {
    UserManager public userManager;
    UserManagerOp public userManagerOp;
    AssetManager public assetManager;

    address public constant ADMIN = address(1);
    address public constant MEMBER = address(2);
    address public constant ACCOUNT = address(3);
    address public constant SIMO = address(4);
    uint256 public constant maxOverdue = 1000;
    uint256 public constant effectiveCount = 3;
    uint256 public constant maxVouchers = 500;
    uint256 public constant maxVouchees = 1000;

    function setUp() public virtual {
        address userManagerLogic = address(new UserManager());
        address userManagerOpLogic = address(new UserManagerOp());

        deployMocks();

        userManager = UserManager(
            deployProxy(
                userManagerLogic,
                abi.encodeWithSignature(
                    "__UserManager_init(address,address,address,address,address,uint256,uint256,uint256,uint256)",
                    address(assetManagerMock),
                    address(unionTokenMock),
                    address(erc20Mock),
                    address(comptrollerMock),
                    ADMIN,
                    maxOverdue,
                    effectiveCount,
                    maxVouchers,
                    maxVouchees
                )
            )
        );

        userManagerOp = UserManagerOp(
            deployProxy(
                userManagerOpLogic,
                abi.encodeWithSignature(
                    "__UserManager_init(address,address,address,address,address,uint256,uint256,uint256,uint256)",
                    address(assetManagerMock),
                    address(unionTokenMock),
                    address(erc20Mock),
                    address(comptrollerMock),
                    ADMIN,
                    maxOverdue,
                    effectiveCount,
                    maxVouchers,
                    maxVouchees
                )
            )
        );

        AssetManager assetManagerLogic = new AssetManager();
        assetManager = AssetManager(
            deployProxy(
                address(assetManagerLogic),
                abi.encodeWithSignature("__AssetManager_init(address,address)", [ADMIN, address(marketRegistryMock)])
            )
        );

        vm.startPrank(ADMIN);
        userManager.setUToken(address(uTokenMock));
        userManager.addMember(MEMBER);
        userManagerOp.setUToken(address(uTokenMock));
        assetManager.addToken(address(erc20Mock));
        assetManager.addAdapter(address(adapterMock));

        vm.stopPrank();

        erc20Mock.mint(MEMBER, 100 ether);
        erc20Mock.mint(SIMO, 100 ether);
        erc20Mock.mint(address(this), 100 ether);

        erc20Mock.approve(address(userManager), type(uint256).max);

        vm.startPrank(MEMBER);
        erc20Mock.approve(address(userManager), type(uint256).max);
        erc20Mock.approve(address(userManagerOp), type(uint256).max);
        vm.stopPrank();

        vm.startPrank(SIMO);
        erc20Mock.approve(address(userManager), type(uint256).max);
        erc20Mock.approve(address(userManagerOp), type(uint256).max);
        vm.stopPrank();
    }
}

```

Add to TestStake.t.sol this function:

```solidity
  function testStakerGetsLessRewardIfPreviousDepositFailed() public {
        uint96 stake1Amount = 10e18;
        uint96 stake2Amount = 5e18;

        //1st succesfull stake of 10e18

        vm.startPrank(SIMO);

        userManager.stake(stake1Amount);
        uint256 stakedAmount1 = userManager.getStakerBalance(SIMO);
        assertEq(stakedAmount1, stake1Amount);
        uint256 totalStaked1 = userManager.totalStaked();
        uint256 totalFrozen = userManager.totalFrozen();
        assertEq(totalFrozen, 0);
        assertEq(totalStaked1 - totalFrozen, stake1Amount);

        vm.stopPrank();

        vm.startPrank(ADMIN);

        //Remove ERC20Mock to make isMarketSupported return false
        //The admin will do this if for example, aave decided to disable the market for whatever reasons

        assetManager.removeToken(address(erc20Mock));

        vm.warp(1 days);

        //2nd stake where the market is no supported

        vm.stopPrank();

        vm.startPrank(SIMO);

        //fail deposit which returns true

        userManager.stake(stake2Amount);

        vm.startPrank(ADMIN);

        //make market supported again

        assetManager.addToken(address(erc20Mock));

        vm.stopPrank();

        vm.startPrank(SIMO);

        //SIMO balance before unstake

        assertEq(erc20Mock.balanceOf(address(SIMO)), 85e18); //100 ether minted in the setup - 15

        //call unstake with 10e18, which should unstake 15e18

        userManager.unstake(stake1Amount);

        //SIMO balance before unstake

        assertEq(erc20Mock.balanceOf(address(SIMO)), 95e18); //balance now is +10e18 instead of +15e18
    }
```

Run `forge test --match-test testStakerGetsLessRewardIfPreviousDepositFailed -vvvv`

**Comment by WangSecurity** [comment]
@cholakovvv thank you for that test!

I modified the last part to the following:

```solidity
        userManager.unstake(stake1Amount + stake2Amount);

        //SIMO balance before unstake

        assertEq(erc20Mock.balanceOf(address(SIMO)), 100e18); //balance now is +10e18 instead of +15e18
```

Basically, I tested if the user can unstake 15e18, i.e. the entire staked amount, and the last line checked that their balance is 100e18, i.e. their initial balance. The test works, so the problem is not in the code but because the user called `unstake` with 10e18, if it's called with 15e18 it still works perfectly and allows them to get their staked amount back.

If I'm not missing something, planning to reject the escalation and leave the issue as it is, since there's no issue.
